<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

<h1 align=center><font size = 5>Assignment</font></h1>


# Juan Alberto Martínez Cadena

# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a Db2 database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:
[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:
[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:
[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. Click on the links below to download and save the datasets (.CSV files):

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Chicago Census Data</a>

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Chicago Public Schools</a>

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Chicago Crime Data</a>

**NOTE:** Ensure you have downloaded the datasets using the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables



##### Open the Db2 console, open the LOAD tool, Select / Drag the .CSV file. Name the new tables as follows:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


### Connect to the database

Let us first load the SQL extension and establish a connection with the database


In [1]:
!pip install sqlalchemy

!pip install sqlalchemy==1.3.9

!pip install ibm_db_sa

!pip install ipython-sql

     |████████████████████████████████| 6.0 MB 7.6 MB/s eta 0:00:01 | 716 kB 7.6 MB/s eta 0:00:01
  Created wheel for sqlalchemy: filename=SQLAlchemy-1.3.9-cp38-cp38-linux_x86_64.whl size=1209485 sha256=0140551a9c44a08bb4b31b679af90fd42b4a91a02d1d69d5e42934ecff42188a
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/cb/43/46/fa638f2422554332b7865d600275b24568bf60e76104a94bb4
Successfully built sqlalchemy
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.4.22
    Uninstalling SQLAlchemy-1.4.22:
      Successfully uninstalled SQLAlchemy-1.4.22
     |████████████████████████████████| 42 kB 2.1 MB/s  eta 0:00:01
  Created wheel for prettytable: filename=prettytable-0.7.2-py3-none-any.whl size=13700 sha256=450e9ebeee0d95ff96dc0e0d5f21fee972cb61db66ebaabe0693096f8ade9663
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/48/6d/77/9517cb933af254f51a446f1a5ec9c2be3e45f17384940bce68
Successfully built prettytable


In [2]:
%load_ext sql

In the next cell enter your db2 connection string. Recall you created Service Credentials for your Db2 instance in first lab in Week 3. 

In [3]:
%sql ibm_db_sa://bcf21776:CmoLnxH9E7oRYy1b@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB?security=SSL

In [4]:
%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES where TABSCHEMA='BCF21776'

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('BCF21776', 'TEST', datetime.datetime(2021, 9, 20, 19, 52, 54, 535109)),
 ('BCF21776', 'PETSALE', datetime.datetime(2021, 9, 20, 20, 1, 53, 147898)),
 ('BCF21776', 'EMPLOYEES', datetime.datetime(2021, 9, 21, 4, 43, 31, 575713)),
 ('BCF21776', 'JOB_HISTORY', datetime.datetime(2021, 9, 21, 4, 43, 31, 749572)),
 ('BCF21776', 'JOBS', datetime.datetime(2021, 9, 21, 4, 43, 31, 930138)),
 ('BCF21776', 'DEPARTMENTS', datetime.datetime(2021, 9, 21, 4, 43, 32, 102504)),
 ('BCF21776', 'LOCATIONS', datetime.datetime(2021, 9, 21, 4, 43, 32, 287870)),
 ('BCF21776', 'PETRESCUE', datetime.datetime(2021, 9, 21, 19, 37, 15, 334905)),
 ('BCF21776', 'TEST2', datetime.datetime(2021, 9, 24, 20, 9, 1, 292560)),
 ('BCF21776', 'COUTRY', datetime.datetime(2021, 9, 24, 20, 13, 24, 482858)),
 ('BCF21776', 'INSTRUCTOR', datetime.datetime(2021, 9, 25, 22, 38, 41, 398230)),
 ('BCF21776', 'INTERNATIONAL_STUDENT_TEST_SCORES', datetime.datetime(2021, 9, 26, 5, 12, 39, 117427)),
 ('BCF21776', 'CHICAGO_SOCIOECONOMIC_DATA', datetime.datetime(2021, 9, 26, 5, 20, 2, 430841)),
 ('BCF21776', 'INSTRUCTOR2', datetime.datetime(2021, 9, 28, 0, 50, 26, 269720)),
 ('BCF21776', 'STUDENT_TEST_SCORES', datetime.datetime(2021, 9, 28, 0, 54, 41, 462559)),
 ('BCF21776', 'STUDENT_TEST_SCORES2', datetime.datetime(2021, 9, 28, 0, 57, 33, 188165)),
 ('BCF21776', 'SCHOOLS', datetime.datetime(2021, 9, 28, 7, 4, 24, 488945)),
 ('BCF21776', 'CENSUS_DATA', datetime.datetime(2021, 9, 28, 8, 13, 4, 668806)),
 ('BCF21776', 'CHICAGO_PUBLIC_SCHOOLS', datetime.datetime(2021, 9, 28, 9, 12, 58, 183004)),
 ('BCF21776', 'CHICAGO_CRIME_DATA', datetime.datetime(2021, 9, 28, 9, 20, 10, 785612))]

## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [5]:
%sql SELECT COUNT(*) FROM CHICAGO_CRIME_DATA;

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[(533,)]

### Problem 2

##### List community areas with per capita income less than 11000.


In [6]:
%sql SELECT COMMUNITY_AREA_NAME, PER_CAPITA_INCOME FROM CENSUS_DATA WHERE PER_CAPITA_INCOME < 11000;

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('West Garfield Park', '10934'),
 ('South Lawndale', '10402'),
 ('Fuller Park', '10432'),
 ('Riverdale', '8201')]

### Problem 3

##### List all case numbers for crimes  involving minors?(children are not considered minors for the purposes of crime analysis)


In [7]:
%%sql SELECT CASE_NUMBER , PRIMARY_TYPE 
from CHICAGO_CRIME_DATA 
where DESCRIPTION like '%MINOR%'

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('HL266884', 'LIQUOR LAW VIOLATION'), ('HK238408', 'LIQUOR LAW VIOLATION')]

### Problem 4

##### List all kidnapping crimes involving a child?


In [8]:
%%sql SELECT  CASE_NUMBER, PRIMARY_TYPE, DESCRIPTION 
from CHICAGO_CRIME_DATA 
where PRIMARY_TYPE='KIDNAPPING' 

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('HN144152', 'KIDNAPPING', 'CHILD ABDUCTION/STRANGER')]

### Problem 5

##### What kinds of crimes were recorded at schools?


In [9]:
%%sql select DISTINCT(PRIMARY_TYPE), LOCATION_DESCRIPTION, DESCRIPTION
from CHICAGO_CRIME_DATA
where LOCATION_DESCRIPTION like '%SCHOOL%'

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('PUBLIC PEACE VIOLATION', 'SCHOOL, PRIVATE, BUILDING', 'BOMB THREAT'),
 ('PUBLIC PEACE VIOLATION', 'SCHOOL, PUBLIC, BUILDING', 'BOMB THREAT'),
 ('NARCOTICS', 'SCHOOL, PUBLIC, BUILDING', 'MANU/DEL:CANNABIS 10GM OR LESS'),
 ('NARCOTICS', 'SCHOOL, PUBLIC, GROUNDS', 'POSS: HEROIN(WHITE)'),
 ('BATTERY', 'SCHOOL, PUBLIC, BUILDING', 'PRO EMP HANDS NO/MIN INJURY'),
 ('ASSAULT', 'SCHOOL, PUBLIC, GROUNDS', 'PRO EMP HANDS NO/MIN INJURY'),
 ('BATTERY', 'SCHOOL, PUBLIC, BUILDING', 'SIMPLE'),
 ('BATTERY', 'SCHOOL, PUBLIC, GROUNDS', 'SIMPLE'),
 ('CRIMINAL TRESPASS', 'SCHOOL, PUBLIC, GROUNDS', 'TO LAND'),
 ('CRIMINAL DAMAGE', 'SCHOOL, PUBLIC, GROUNDS', 'TO VEHICLE')]

### Problem 6

##### List the average safety score for all types of schools.


In [10]:
%%sql SELECT AVG(SAFETY_SCORE) as AVERANGE_SAFETY_SCORE
from CHICAGO_PUBLIC_SCHOOLS

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('49.50487329434697855750487329434698',)]

In [11]:
%sql SELECT NAME_OF_SCHOOL, SAFETY_SCORE FROM CHICAGO_PUBLIC_SCHOOLS WHERE SAFETY_SCORE=49 or SAFETY_SCORE=50;

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('Air Force Academy High School', '49'),
 ('Alexander Graham Elementary School', '49'),
 ('Bernhard Moos Elementary School', '50'),
 ('Daniel Hale Williams Prep School of Medicine', '50'),
 ('Edward Jenner Elementary Academy of the Arts', '50'),
 ('Francisco I Madero Middle School', '49'),
 ('George B McClellan Elementary School', '49'),
 ('Graeme Stewart Elementary School', '49'),
 ('Ida B Wells Preparatory Elementary Academy', '49'),
 ('James Ward Elementary School', '49'),
 ('John C Burroughs Elementary School', '50'),
 ('Joseph Brennemann Elementary School', '50'),
 ('Josiah Pickard Elementary School', '50'),
 ('Louis Pasteur Elementary School', '50'),
 ('Marcus Moziah Garvey Elementary School', '50'),
 ('Mason High School', '50'),
 ('Matthew Gallistel Elementary Language Academy', '49'),
 ('Nathan R Goldblatt Elementary School', '49'),
 ('Near North Elementary School', '50'),
 ('Uplift Community High School', '50'),
 ('Victor Herbert Elementary School', '50'),
 ('William Bishop Owen Scholastic Academy Elementary ', '49'),
 ('Williams Preparatory Academy Middle School', '49')]

### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [12]:
%%sql select COMMUNITY_AREA_NAME, PERCENT_HOUSEHOLDS_BELOW_POVERTY 
from CENSUS_DATA order by PERCENT_HOUSEHOLDS_BELOW_POVERTY limit 5

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('Ashburn', '10.4'),
 ('Dunning', '10.6'),
 ('Lincoln Square', '10.9'),
 ('Lake View', '11.4'),
 ('Calumet Heights', '11.5')]

### Problem 8

##### Which community area is most crime prone?


In [13]:
%%sql select COMMUNITY_AREA_NUMBER, COUNT(COMMUNITY_AREA_NUMBER) AS FREQUENCY
from CHICAGO_CRIME_DATA 
group by COMMUNITY_AREA_NUMBER
order by COUNT(COMMUNITY_AREA_NUMBER) desc limit 1

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('25', 43)]

Double-click **here** for a hint

<!--
Query for the 'community area number' that is most crime prone.
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [14]:
%%sql select COMMUNITY_AREA_NAME, HARDSHIP_INDEX 
from CENSUS_DATA where HARDSHIP_INDEX = (select MAX(HARDSHIP_INDEX) from CENSUS_DATA)

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('Riverdale', '98')]

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [15]:
%%sql select COMMUNITY_AREA_NAME 
from CENSUS_DATA 
where COMMUNITY_AREA_NUMBER = (
    SELECT CCD.COMMUNITY_AREA_NUMBER
    from CHICAGO_CRIME_DATA AS CCD
    group by CCD.COMMUNITY_AREA_NUMBER
    order by COUNT(CCD.COMMUNITY_AREA_NUMBER) desc limit 1)

 * ibm_db_sa://bcf21776:***@fbd88901-ebdb-4a4f-a32e-9822b9fb237b.c1ogj3sd0tgtu0lqde00.databases.appdomain.cloud:32731/BLUDB
Done.


[('Austin',)]

Copyright © 2020 [cognitiveclass.ai](cognitiveclass.ai?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).


## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>

## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
